File: ios/CreatePrivateDealerOS/CreatePrivateDealerOS/ViewModels/ConversationsViewModel.swift
import Foundation

@MainActor
final class ConversationsViewModel: ObservableObject {
    @Published var conversations: [Conversation] = []
    @Published var searchText = ""
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let repository: DealerOSRepository

    init(repository: DealerOSRepository) {
        self.repository = repository
    }

    var filteredConversations: [Conversation] {
        guard !searchText.isEmpty else {
            return conversations
        }

        let query = searchText.lowercased()
        return conversations.filter { conversation in
            let leadName = conversation.lead?.displayName.lowercased() ?? ""
            let preview = conversation.lastMessagePreview?.lowercased() ?? ""
            return leadName.contains(query) || preview.contains(query)
        }
    }

    func load(session: AppSession) async {
        guard !isLoading else { return }
        isLoading = true
        defer { isLoading = false }

        do {
            conversations = try await repository.fetchConversations(session: session)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: flutter/create_private_dealer_os/lib/application/view_models/conversations_view_model.dart
import 'package:flutter/foundation.dart';

import '../../domain/models/app_session.dart';
import '../../domain/models/conversation.dart';
import '../../domain/repositories/dealer_os_repository.dart';

class ConversationsViewModel extends ChangeNotifier {
  ConversationsViewModel({required DealerOsRepository repository})
      : _repository = repository;

  final DealerOsRepository _repository;

  List<Conversation> _conversations = <Conversation>[];
  String _searchText = '';
  bool _isLoading = false;
  String? _errorMessage;

  List<Conversation> get conversations => _conversations;
  String get searchText => _searchText;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  set searchText(String value) {
    _searchText = value;
    notifyListeners();
  }

  List<Conversation> get filteredConversations {
    if (_searchText.trim().isEmpty) {
      return _conversations;
    }

    final String query = _searchText.toLowerCase();
    return _conversations.where((Conversation conversation) {
      final String leadName = (conversation.lead?.displayName ?? '').toLowerCase();
      final String preview = (conversation.lastMessagePreview ?? '').toLowerCase();
      return leadName.contains(query) || preview.contains(query);
    }).toList();
  }

  Future<void> load({required AppSession session}) async {
    if (_isLoading) {
      return;
    }

    _isLoading = true;
    _errorMessage = null;
    notifyListeners();

    try {
      _conversations = await _repository.fetchConversations(session: session);
    } catch (error) {
      _errorMessage = error.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }
}
